### ISSUE TITLE
URIs in context keys

### ISSUE THREAD

**Issue by roblourens** [microsoft/vscode#147732]
We have a 'resource' context key which has the full URI of the open editor. Even though it's a URI type, it's actually used as a string:

https://github.com/microsoft/vscode/blob/204d0247033ab7d2f6d1949ae6fd6d87e46a91b3/src/vs/workbench/contrib/preferences/browser/preferences.contribution.ts#L254

and it only works as a string. The equals expr coerces the real URI to a string:

https://github.com/microsoft/vscode/blob/4a130c40ed876644ed8af2943809d08221375408/src/vs/platform/contextkey/common/contextkey.ts#L457

and the "in" expr, which I want to use, doesn't work with URI objects because it's checking object equality.

So I suggest that we change the type to just be a string that only contains the toStringified URI. Does that make sense or am I missing something about how this is meant to be used?

**Comment by roblourens** [comment]
Basically in https://github.com/microsoft/vscode/issues/146686 I added a `notebookCellResource` context key which is a URI toString and I'm wondering whether to make 'resource' work the same. The alternative is changing some context expression types to know how to check URI equality.

**Comment by roblourens** [comment]
This issue has gotten a little chaotic but here's one proposal: https://github.com/microsoft/vscode/pull/147734

**Comment by roblourens** [comment]
https://github.com/microsoft/vscode/pull/147926 makes it clear that we only want a string value in a context key value. To close this out, I still need a way for an extension to set a URI in a context key but have it stringified on the frontend so that it goes through uri-transformation and can be compared to something like my new `notebookCellResource` key. See https://github.com/microsoft/vscode/issues/146686#issuecomment-1097403109 for a usage example.

So I guess I would propose stringifying it in the setContext command at this point https://github.com/microsoft/vscode/blob/93aad560e1190f81b456b09d0b05bca4382c1033/src/vs/platform/contextkey/browser/contextKeyService.ts#L578 and similar to my original PR, I'm not sure whether we want to special-case URIs or if there is something more generic to do there.

**Comment by alexdima** [comment]
@roblourens I'm super sorry, but with things going in so many directions I'm having trouble understanding what the latest proposal is and the motivation behind it.

If I understand correctly, we have done a big adoption with #147926 and we have now eliminated non-primitive values from the context keys defined by VS Code core. 

`setContext` is exposed as a command for extensions. Coercing the passed in value to a string might cause breakage if the passed in value is read somewhere via the `IContextKeyService.getContextKeyValue` API and passed back as an argument to a command defined by the same extension. We could introduce a warning if the passed in value is not a primitive. 

Would it be possible to identify the problematic extension that calls `setContext` and change the extension code to pass in primitives instead of URIs?

**Comment by roblourens** [comment]
Yeah this set of issues has taken some turns. cc @jrieken. I am still trying to accomplish basically what I described in https://github.com/microsoft/vscode/issues/146686#issuecomment-1097403109. Let me update that for what has happened and summarize here. If it's hard to follow, call me any morning.

- I am trying to give an extension the ability to contribute a menu button to the toolbar for a particular notebook cell. The extension contributes a menu item with `"when": "notebookCellResource in jupyter.runByLineCells"` where `notebookCellResource` is a built-in context key which is the (stringified) URI of a notebook cell
- The extension runs something like
`executeCommand('setContext', 'jupyter.runByLineCells', [cell.uri])` - they do not stringify the URI at this point because it needs to go through the URI transformer so that it will match the uri on the frontend

The problem is, that one `in` operator did not work for URIs. So we had this conversation about whether to support URIs or other objects as context key values, or only primitives, and it seems the intent is that context key values should only be primitives. My original idea was that the 'setContext' command would provide the service of coercing URIs to strings on the frontend so that the scenario above can work. Alternatively, we can still declare URIs to a primitive for the purposes of evaluating context key expressions.

> and passed back as an argument to a command defined by the same extension
How does a context value get passed back to an extension command?

**Comment by jrieken** [comment]
> My original idea was that the 'setContext' command would provide the service of coercing URIs to strings (or maybe any invalid value to a primitive) on the frontend so that the scenario above can work

 I like that better than treating URIs as primitives - either way to need to traverse/massage the values given via 'setContext' because URIs would need "reviving".  

> How does a context value get passed back to an extension command?

Arguments are independent of context key values. When are arguments are provided than that's usually dependent on the menu. E.g the (code) editor context menu passes the URI of the current document, tree view context menus pass the tree item etc. For notebook cells menus it would make sense to pass the cell for which the menu has showing etc.

**Comment by alexdima** [comment]
> The extension runs something like
`executeCommand('setContext', 'jupyter.runByLineCells', [cell.uri])` - they do not stringify the URI at this point because **it needs to go through the URI transformer** so that it will match the uri on the frontend

Aha! Thanks for clarifying that. I agree that URIs need to cross over as URIs to benefit from transforming, so therefore we need to revive them and `.toString()` them [here](https://github.com/microsoft/vscode/blob/93aad560e1190f81b456b09d0b05bca4382c1033/src/vs/platform/contextkey/browser/contextKeyService.ts#L578). I agree that it might be best to just add a special case for URIs.

**Comment by roblourens** [comment]
Looks perfect, thank you!
